File: python/neuralNet.py
import sys
import numpy as np

class neuralNet:
	def __init__(self, sizeInput, sizeHidden, sizeOutput):
		self.learningRate = 0.4
		self.momentum = 0.8

		self.sizeInput = sizeInput
		self.sizeHidden = sizeHidden
		self.sizeOutput = sizeOutput

		# Allocate neurons
		self.neuronsInput = np.zeros((1,sizeInput))
		self.neuronsHidden = np.zeros((1,sizeHidden+1))
		self.neuronsOutput = np.zeros((1,sizeOutput+1))

		# Allocate weights
		self.weightsInputToHidden = np.random.random((sizeInput+1,sizeHidden)) * 0.24 - 0.12
		self.weightsHiddenToOutput = np.random.random((sizeHidden+1,sizeOutput)) * 0.24 - 0.12

		# Allocate training data
		self.errorHidden = np.zeros((1,sizeHidden+1))
		self.errorOutput = np.zeros((1,sizeOutput))
		self.idealOutput = np.zeros((1,sizeOutput))
		self.oldDeltaWeightInputToHidden = np.zeros((sizeInput+1,sizeHidden))
		self.oldDeltaWeightHiddenToOutput = np.zeros((sizeHidden+1,sizeOutput))
		self.newDeltaWeightInputToHidden = np.zeros((sizeInput+1,sizeHidden))
		self.newDeltaWeightHiddenToOutput = np.zeros((sizeHidden+1,sizeOutput))

	def activationFunction(self, in_num):
		# Many functions may be used for activation
		# Sigmoid used by default
		return 1.0 / (1.0 + np.exp(-in_num))

	def feedForward(self, in_arr):
		self.neuronsInput = np.concatenate((in_arr,np.array([[-1]])),axis=1)
		self.neuronsHidden = np.concatenate((self.activationFunction(np.dot(self.neuronsInput,self.weightsInputToHidden)),np.array([[-1]])),axis=1)
		self.neuronsOutput = self.activationFunction(np.dot(self.neuronsHidden,self.weightsHiddenToOutput))

	def saveWeights(self, outFile):
		f = open(outFile, 'w')
		f.write("Dimensions:\n")
		f.write("%d %d %d\n" % (self.sizeInput, self.sizeHidden, self.sizeOutput))

		f.write("Weights Input To Hidden:\n")
		f.writelines(' '.join("%.50f" % j for j in i) + '\n' for i in self.weightsInputToHidden)

		f.write("Weights Hidden To Output:\n")
		f.writelines(' '.join("%.50f" % j for j in i) + '\n' for i in self.weightsHiddenToOutput)
		f.close()
		return True

	def loadWeights(self, inFile):
		f = open(inFile, 'r')
		f.readline() # Dimensions Label
		line = f.readline()
		line = [int(i) for i in line.split()]
		if (line[0] == self.sizeInput) and (line[1] == self.sizeHidden) and (line[2] == self.sizeOutput):
			f.readline() # Weights Label
			weightsIToH = []
			for x in xrange(0,self.sizeInput + 1):
				line = f.readline()
				line = [float(i) for i in line.split()]
				weightsIToH.append(line)

			f.readline() # Weights Label
			weightsHToO = []
			for x in xrange(0,self.sizeHidden + 1):
				line = f.readline()
				line = [float(i) for i in line.split()]
				weightsHToO.append(line)
			self.weightsInputToHidden = np.array(weightsIToH)
			self.weightsHiddenToOutput = np.array(weightsHToO)
			f.close()
			return True
		else:
			return False

	def trainBatch(self, inputs, outputs, maxEpochs):
		numTests = np.size(inputs,0)
		training = int(0.6 * numTests)
		generalizing = int(0.8 * numTests)
		epoch = 1

		while epoch <= 20:
			TMSE = 0
			Tacc = 0
			GMSE = 0
			Gacc = 0
			VMSE = 0
			Vacc = 0
			r = 0
			count = 0
			for i in xrange(0,training):
				# Feed forward and test accuracy
				self.feedForward(inputs[[i],:])

				TMSE = TMSE + np.power(1-self.neuronsOutput[0,outputs[i,0]],2)
				r = np.argmax(self.neuronsOutput)
				if r == outputs[i,0]:
					Tacc = Tacc + 1

				# Back propagation and weight updates
				self.idealOutput = np.zeros((1,self.sizeOutput))
				self.idealOutput[0,outputs[i,0]] = 1
				self.errorOutput = self.neuronsOutput*(1-self.neuronsOutput)*(self.idealOutput-self.neuronsOutput)
				self.errorHidden = self.neuronsHidden*(1-self.neuronsHidden)*sum(np.transpose(self.weightsHiddenToOutput*self.errorOutput))
				self.newDeltaWeightHiddenToOutput = self.learningRate*np.dot(np.transpose(self.neuronsHidden),self.errorOutput) + self.momentum*self.oldDeltaWeightHiddenToOutput
				self.weightsHiddenToOutput = self.weightsHiddenToOutput + self.newDeltaWeightHiddenToOutput
				self.newDeltaWeightInputToHidden = self.learningRate*np.dot(np.transpose(self.neuronsInput),self.errorHidden[[0],0:-1]) + self.momentum*self.oldDeltaWeightInputToHidden
				self.weightsInputToHidden = self.weightsInputToHidden + self.newDeltaWeightInputToHidden

				self.oldDeltaWeightHiddenToOutput = self.newDeltaWeightHiddenToOutput
				self.oldDeltaWeightInputToHidden = self.newDeltaWeightInputToHidden

				count = count + 1
			TMSE = TMSE / count
			Tacc = Tacc / float(count) * 100.0

			count = 0
			for i in xrange(training,generalizing):
				self.feedForward(inputs[[i],:])

				GMSE = GMSE + np.power(1-self.neuronsOutput[0,outputs[i,0]],2)
				r = np.argmax(self.neuronsOutput)
				if r == outputs[i,0]:
					Gacc = Gacc + 1

				count = count + 1
			GMSE = GMSE / count
			Gacc = Gacc / float(count) * 100.0

			print("Epoch: %d, TMSE: %f, Tacc: %f, GMSE: %f, Gacc: %f" % (epoch,TMSE,Tacc,GMSE,Gacc))
			sys.stdout.flush()

			if (((TMSE < 0.05) and (GMSE < 0.05)) or ((Tacc > 95) and (Gacc > 95))):
				break
			epoch = epoch + 1

		count = 0
		for i in xrange(generalizing,numTests):
			self.feedForward(inputs[[i],:])

			VMSE = VMSE + np.power(1-self.neuronsOutput[0,outputs[i,0]],2)
			r = np.argmax(self.neuronsOutput)
			if r == outputs[i,0]:
				Vacc = Vacc + 1

			count = count + 1
		VMSE = VMSE / count
		Vacc = Vacc / float(count) * 100.0

		print("VMSE: %f, Vacc: %f" % (VMSE, Vacc))

	def trainLive(self, input, output):
		# Feed forward
		self.feedForward(input)

		# Back propagation and weight updates
		self.idealOutput = np.zeros((1,self.sizeOutput))
		self.idealOutput[0,output] = 1
		self.errorOutput = self.neuronsOutput*(1-self.neuronsOutput)*(self.idealOutput-self.neuronsOutput)
		self.errorHidden = self.neuronsHidden*(1-self.neuronsHidden)*sum(np.transpose(self.weightsHiddenToOutput*self.errorOutput))
		self.newDeltaWeightHiddenToOutput = self.learningRate*np.dot(np.transpose(self.neuronsHidden),self.errorOutput) + self.momentum*self.oldDeltaWeightHiddenToOutput
		self.weightsHiddenToOutput = self.weightsHiddenToOutput + self.newDeltaWeightHiddenToOutput
		self.newDeltaWeightInputToHidden = self.learningRate*np.dot(np.transpose(self.neuronsInput),self.errorHidden[[0],0:-1]) + self.momentum*self.oldDeltaWeightInputToHidden
		self.weightsInputToHidden = self.weightsInputToHidden + self.newDeltaWeightInputToHidden

		self.oldDeltaWeightHiddenToOutput = self.newDeltaWeightHiddenToOutput
		self.oldDeltaWeightInputToHidden = self.newDeltaWeightInputToHidden

		return self.neuronsOutput

	def classify(self,input):
		self.feedForward(input)
		return np.argmax(self.neuronsOutput)

File: cpp/neuralNet.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <math.h>
#include <cstdlib>

#include "neuralNet.h"

using namespace std;

// Constructor
neuralNet::neuralNet(int numIn, int numHid, int numOut) : sizeInput(numIn), sizeHidden(numHid), sizeOutput(numOut)
{
	learningRate = 0.4;
	momentum = 0.8;

	// Allocate neurons
	neuronsInput = new double[sizeInput + 1];
	neuronsHidden = new double[sizeHidden + 1];
	neuronsOutput = new double[sizeOutput];

	// Allocate weights
	weightsInputToHidden = new double*[sizeInput + 1];
	weightsHiddenToOutput = new double*[sizeHidden + 1];

	// Allocate training data
	errorHidden = new double[sizeHidden + 1];
	errorOutput = new double[sizeOutput];
	idealOutput = new double[sizeOutput];
	oldDeltaWeightInputToHidden = new double*[sizeInput + 1];
	oldDeltaWeightHiddenToOutput = new double*[sizeHidden + 1];
	newDeltaWeightInputToHidden = new double*[sizeInput + 1];
	newDeltaWeightHiddenToOutput = new double*[sizeHidden + 1];

	// Initialize layers
	for (int i = 0; i <= sizeInput; i++)
	{
		neuronsInput[i] = 0;
		weightsInputToHidden[i] = new double[sizeHidden];
		oldDeltaWeightInputToHidden[i] = new double[sizeHidden];
		newDeltaWeightInputToHidden[i] = new double[sizeHidden];
		for (int j = 0; j < sizeHidden; j++)
		{
			weightsInputToHidden[i][j] = 0;
			oldDeltaWeightInputToHidden[i][j] = 0;
			newDeltaWeightInputToHidden[i][j] = 0;
		}
	}
	neuronsInput[sizeInput] = -1; // Bias neuron
	for (int j = 0; j <= sizeHidden; j++)
	{
		neuronsHidden[j] = 0;
		errorHidden[j] = 0;
		weightsHiddenToOutput[j] = new double[sizeOutput];
		oldDeltaWeightHiddenToOutput[j] = new double[sizeOutput];
		newDeltaWeightHiddenToOutput[j] = new double[sizeOutput];
		for (int k = 0; k < sizeOutput; k++)
		{
			weightsHiddenToOutput[j][k] = 0;
			oldDeltaWeightHiddenToOutput[j][k] = 0;
			newDeltaWeightHiddenToOutput[j][k] = 0;
		}
	}
	neuronsHidden[sizeHidden] = -1; // Bias neuron
	for (int k = 0; k < sizeOutput; k++)
	{
		neuronsOutput[k] = 0;
		errorOutput[k] = 0;
		idealOutput[k] = 0;
	}

	initializeWeights();
}

// Destructor
neuralNet::~neuralNet()
{
	for (int i = 0; i <= sizeInput; i++)
	{
		delete[] weightsInputToHidden[i];
		delete[] oldDeltaWeightInputToHidden[i];
		delete[] newDeltaWeightInputToHidden[i];
	}
	for (int j = 0; j <= sizeHidden; j++)
	{
		delete[] weightsHiddenToOutput[j];
		delete[] oldDeltaWeightHiddenToOutput[j];
		delete[] newDeltaWeightHiddenToOutput[j];
	}
	delete[] weightsInputToHidden;
	delete[] oldDeltaWeightInputToHidden;
	delete[] newDeltaWeightInputToHidden;
	delete[] weightsHiddenToOutput;
	delete[] oldDeltaWeightHiddenToOutput;
	delete[] newDeltaWeightHiddenToOutput;

	delete[] errorHidden;
	delete[] errorOutput;
	delete[] idealOutput;

	delete[] neuronsInput;
	delete[] neuronsHidden;
	delete[] neuronsOutput;
}

void neuralNet::initializeWeights()
{
	// Range of weights
	double rangeHidden = 1/sqrt((double)sizeInput);
	double rangeOutput = 1/sqrt((double)sizeHidden);

	// Randomly set weights
	for (int i = 0; i <= sizeInput; i++)
	{
		for (int j = 0; j < sizeHidden; j++)
			weightsInputToHidden[i][j] = (((double)(rand() % 100 + 1))/100.0) * 2.0 * rangeHidden - rangeHidden;
	}
	for (int j = 0; j <= sizeHidden; j++)
	{
		for (int k = 0; k < sizeOutput; k++)
			weightsHiddenToOutput[j][k] = (((double)(rand() % 100 + 1))/100.0) * 2.0 * rangeOutput - rangeOutput;
	}
}

double neuralNet::activationFunction(double in)
{
	// Many functions may be used for activation
	// Sigmoid used by default
	return 1.0 / (1.0 + exp(-in));
}

void neuralNet::feedForward(double* in)
{
	for (int i = 0; i < sizeInput; i++)
		neuronsInput[i] = in[i];
	for (int j = 0; j < sizeHidden; j++)
	{
		neuronsHidden[j] = 0;
		for (int i = 0; i <= sizeInput; i++)
			neuronsHidden[j] += neuronsInput[i] * weightsInputToHidden[i][j];
		neuronsHidden[j] = activationFunction(neuronsHidden[j]);
	}
	for (int k = 0; k < sizeOutput; k++)
	{
		neuronsOutput[k] = 0;
		for (int j = 0; j <= sizeHidden; j++)
			neuronsOutput[k] += neuronsHidden[j] * weightsHiddenToOutput[j][k];
		neuronsOutput[k] = activationFunction(neuronsOutput[k]);
	}
}

void neuralNet::setLearningRate(double lr)
{
	learningRate = lr;
}

void neuralNet::setMomentum(double m)
{
	momentum = m;
}

double neuralNet::getLearningRate()
{
	return learningRate;
}

double neuralNet::getMomentum()
{
	return momentum;
}

bool neuralNet::saveWeights(char* outFile)
{
	fstream output;
	output.open(outFile, ios::out);

	if (output.is_open())
	{
		output.precision(50);

		output << "Dimensions:\n" << sizeInput << " " << sizeHidden << " " << sizeOutput << endl;

		output << "Weights Input To Hidden:\n";
		for (int i = 0; i <= sizeInput; i++)
		{
			for (int j = 0; j < sizeHidden; j++)
				output << weightsInputToHidden[i][j] << " ";
			output << endl;
		}

		output << "Weights Hidden To Output:\n";
		for (int j = 0; j <= sizeHidden; j++)
		{
			for (int k = 0; k < sizeOutput; k++)
				output << weightsHiddenToOutput[j][k] << " ";
			output << endl;
		}

		output.close();
		return true;
	}
	else
	{
		return false;
	}
}

bool neuralNet::loadWeights(char* inFile)
{
	fstream input;
	input.open(inFile, ios::in);

	if (input.is_open())
	{
		int numInput, numHidden, numOutput;
		string line = "";

		getline(input, line); // Dimensions Label
		input >> numInput >> numHidden >> numOutput;
		getline(input, line); // Clear line feed and newline characters

		if ((numInput == sizeInput) && (numHidden == sizeHidden) && (numOutput == sizeOutput))
		{
			getline(input, line); // Weights Label
			for (int i = 0; i <= sizeInput; i++)
			{
				for (int j = 0; j < sizeHidden; j++)
					input >> weightsInputToHidden[i][j];
				getline(input, line); // Clear line feed and newline characters
			}

			getline(input, line); // Weights Label
			for (int j = 0; j <= sizeHidden; j++)
			{
				for (int k = 0; k < sizeOutput; k++)
					input >> weightsHiddenToOutput[j][k];
				getline(input, line); // Clear line feed and newline characters
			}
			input.close();
			return true;
		}
		else
		{
			input.close();
			return false;
		}
	}
	else
	{
		return false;
	}
}

void neuralNet::trainBatch(double** inputs, int* outputs, int numTests, int maxEpochs)
{
	int training = 0.6 * numTests;
	int generalizing = 0.8 * numTests;

	double TMSE, Tacc, GMSE, Gacc, VMSE, Vacc;
	double max, sum;
	int result, count;

	for (int epoch = 1; epoch <= maxEpochs; epoch++)
	{
		TMSE = 0;
		Tacc = 0;
		GMSE = 0;
		Gacc = 0;
		// Train on 60% of the data
		count = 0;
		for (int test = 0; test < training; test++)
		{
			max = -1.0;
			result = 0;

			// Feed forward and test accuracy
			feedForward(inputs[test]);

			TMSE += pow((1.0 - neuronsOutput[outputs[test]]),2.0);
			for (int k = 0; k < sizeOutput; k++)
			{
				idealOutput[k] = 0;
				if (neuronsOutput[k] > max)
				{
					max = neuronsOutput[k];
					result = k;
				}
			}
			if (result == outputs[test])
				Tacc += 1.0;

			// Back propagation
			idealOutput[outputs[test]] = 1;
			for (int k = 0; k < sizeOutput; k++)
			{
				errorOutput[k] = neuronsOutput[k]*(1-neuronsOutput[k])*(idealOutput[k]-neuronsOutput[k]);
			}
			for (int j = 0; j <= sizeHidden; j++)
			{
				sum = 0;
				for (int k = 0; k < sizeOutput; k++)
				{
					sum += weightsHiddenToOutput[j][k]*errorOutput[k];
					newDeltaWeightHiddenToOutput[j][k] = (learningRate*neuronsHidden[j]*errorOutput[k]) + momentum*oldDeltaWeightHiddenToOutput[j][k];
					oldDeltaWeightHiddenToOutput[j][k] = newDeltaWeightHiddenToOutput[j][k];
					weightsHiddenToOutput[j][k] += newDeltaWeightHiddenToOutput[j][k];
				}
				errorHidden[j] = neuronsHidden[j]*(1-neuronsHidden[j])*sum;
			}
			for (int i = 0; i <= sizeInput; i++)
			{
				for (int j = 0; j < sizeHidden; j++)
				{
					newDeltaWeightInputToHidden[i][j] = (learningRate*neuronsInput[i]*errorHidden[j]) + momentum*oldDeltaWeightInputToHidden[i][j];
					oldDeltaWeightInputToHidden[i][j] = newDeltaWeightInputToHidden[i][j];
					weightsInputToHidden[i][j] += newDeltaWeightInputToHidden[i][j];
				}
			}
			count++;
		}
		TMSE = TMSE / (double)count;
		Tacc = Tacc / (double)count * 100.0;

		// Generalization testing on 20% of the data
		count = 0;
		for (int test = training; test < generalizing; test++)
		{
			max = -1.0;
			result = 0;

			feedForward(inputs[test]);

			GMSE += pow((1 - neuronsOutput[outputs[test]]),2);
			for (int k = 0; k < sizeOutput; k++)
			{
				if (neuronsOutput[k] > max)
				{
					max = neuronsOutput[k];
					result = k;
				}
			}
			if (result == outputs[test])
				Gacc += 1.0;
			count++;
		}
		GMSE = GMSE / (double)count;
		Gacc = Gacc / (double)count * 100.0;

		cout<<"Epoch: "<<epoch<<", TMSE: "<<TMSE<<", Tacc: "<<Tacc<<", GMSE: "<<GMSE<<", Gacc: "<<Gacc<<endl;
		if (((TMSE < 0.05)&&(GMSE < 0.05)) || ((Tacc > 95)&&(Gacc > 95)))
			break;
	}

	// Validating results
	count = 0;
	for (int test = generalizing; test < numTests; test++)
	{
		max = -1.0;
		result = 0;

		feedForward(inputs[test]);

		VMSE += pow((1 - neuronsOutput[outputs[test]]),2);
		for (int k = 0; k < sizeOutput; k++)
		{
			if (neuronsOutput[k] > max)
			{
				max = neuronsOutput[k];
				result = k;
			}
		}
		if (result == outputs[test])
			Vacc += 1.0;
		count++;
	}
	VMSE = VMSE / (double)count;
	Vacc = Vacc / (double)count * 100.0;

	cout << "VMSE: " << VMSE << ", Vacc: " << Vacc << endl;
	return;
}

double* neuralNet::trainLive(double* in, int out)
{
	double sum;

	// Feed forward
	feedForward(in);

	for (int k = 0; k < sizeOutput; k++)
		idealOutput[k] = 0;

	// Back propagation
	idealOutput[out] = 1;
	for (int k = 0; k < sizeOutput; k++)
	{
		errorOutput[k] = neuronsOutput[k]*(1-neuronsOutput[k])*(idealOutput[k]-neuronsOutput[k]);
	}
	for (int j = 0; j <= sizeHidden; j++)
	{
		sum = 0;
		for (int k = 0; k < sizeOutput; k++)
		{
			sum += weightsHiddenToOutput[j][k]*errorOutput[k];
			newDeltaWeightHiddenToOutput[j][k] = (learningRate*neuronsHidden[j]*errorOutput[k]) + momentum*oldDeltaWeightHiddenToOutput[j][k];
			oldDeltaWeightHiddenToOutput[j][k] = newDeltaWeightHiddenToOutput[j][k];
			weightsHiddenToOutput[j][k] += newDeltaWeightHiddenToOutput[j][k];
		}
		errorHidden[j] = neuronsHidden[j]*(1-neuronsHidden[j])*sum;
	}
	for (int i = 0; i <= sizeInput; i++)
	{
		for (int j = 0; j < sizeHidden; j++)
		{
			newDeltaWeightInputToHidden[i][j] = (learningRate*neuronsInput[i]*errorHidden[j]) + momentum*oldDeltaWeightInputToHidden[i][j];
			oldDeltaWeightInputToHidden[i][j] = newDeltaWeightInputToHidden[i][j];
			weightsInputToHidden[i][j] += newDeltaWeightInputToHidden[i][j];
		}
	}
	return neuronsOutput;
}

int neuralNet::classify(double* in)
{
	double max = -1.0;
	int result = 0;

	feedForward(in);

	for (int k = 0; k < sizeOutput; k++)
	{
		if (neuronsOutput[k] > max)
		{
			max = neuronsOutput[k];
			result = k;
		}
	}

	return result;
}
